### PR TITLE
feat: add an option to load mini app by url (MINI-1981)

### PR DESCRIPTION
**Pull request by imaginaris**
# Description
Search bar in sample app now accepts localhost and https urls containing a mini app.
No data is persisted for such mini apps.

## Links
MINI-1981

# Checklist
- [X] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [x] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `fastlane ci` without errors

### CODE REVIEW

**Review by Climbatize — APPROVED** [review]
Mostly good to me, please read my comments

File: MiniApp/Classes/MiniApp.swift
Comment (corycaywood) [inline MiniApp/Classes/MiniApp.swift:55]: For now the ability to load from a URL is supposed to be only used by our Demo App, so we don't really want to support clients using it in their own App yet.

Is there any way to hide a function from the generated documentation? Or restrict it's usage to the example app? For example, on Android we can hide it from the docs using `@suppress` or use `@RestrictTo` to limit it's usage to a specific package. 

If not, probably it's okay. We can just leave out this new function from the user guide so we aren't advertising that it exists.

File: Example/Controllers/DisplayController.swift
Comment (imaginaris) [reply Example/Controllers/DisplayController.swift:16]: Because no MiniAppInfo is created for url loaded miniapp (no id, title etc)

File: Example/Controllers/Extensions/StringExtensions.swift
Comment (imaginaris) [reply Example/Controllers/Extensions/StringExtensions.swift:1]: Oh ok. I didn't notice. I'll move that code then

File: Example/Controllers/Storyboards/Base.lproj/Main.storyboard
Comment (imaginaris) [reply Example/Controllers/Storyboards/Base.lproj/Main.storyboard:809]: I think now is the moment. What do you have in mind as "another UI"?

File: Example/Controllers/ViewController.swift
Comment (imaginaris) [reply Example/Controllers/ViewController.swift:57]: In normal case that's true. Since url loaded miniapp doesn't produce miniAppInfo content, the guard would block the display. I guess I can leave that check for miniapps from the list and add an exception for url loaded ones

File: Example/Info.plist
Comment (imaginaris) [reply Example/Info.plist:64]: yeah... unfortunately that's going to be necessary

File: MiniApp/Classes/Display/MiniAppExternalWebViewController.swift
Comment (imaginaris) [reply MiniApp/Classes/Display/MiniAppExternalWebViewController.swift:12]: Because I couldn't think of any better way to inject this information :/ (without modifying public api)

File: MiniApp/Classes/Display/MiniAppWebView.swift
Comment (imaginaris) [reply MiniApp/Classes/Display/MiniAppWebView.swift:38]: haha :D sorry I think I was meddling with the code around here and it's just a leftover ;) (of course that would be unnecessary refactoring)

File: MiniApp/Classes/MiniApp.swift
Comment (Climbatize) [reply MiniApp/Classes/MiniApp.swift:55]: Unfortunately we don't have this in Swift. The most elegant thing that comes to my mind about this would be to create a subspec dedicated to this feature, `MiniApp/Sample` (https://guides.cocoapods.org/syntax/podspec.html#subspec). I can arrange this if you want

Comment (imaginaris) [reply MiniApp/Classes/MiniApp.swift:55]: Subspec sounds like a good idea. It's either this or removal of documentation for that new method. Which one do you gyus think would be the best solution?

Comment (corycaywood) [reply MiniApp/Classes/MiniApp.swift:55]: Hmm, I think it's probably okay to leave it as is. Eventually we might want to open up this method for general use anyways.

Maybe update the doc for this method to say it's for previewing a mini app? i.e. `This should only be used for previewing a mini app from a local server.`

File: Example/Controllers/Extensions/StringExtensions.swift
Comment (imaginaris) [reply Example/Controllers/Extensions/StringExtensions.swift:1]: I moved that and I realized that I'd have to make that extension public because that `hasHTTPPrefix` is only used in Sample app target

**Review by ts-leojoseph-r — COMMENTED** [review]
@imaginaris Looks really good 👍 , just a few comments.

File: MiniApp/Classes/MiniApp.swift
Comment (ts-leojoseph-r) [inline MiniApp/Classes/MiniApp.swift:109]: Can we have something like a completionHandler or anything that would let the host app know if it fails to load the URL.

File: Example/Controllers/Extensions/StringExtensions.swift
Comment (imaginaris) [reply Example/Controllers/Extensions/StringExtensions.swift:1]: Oh there's `String+MiniApp` and `MiniApp+String` and only the latter is inside the SDK...

File: Example/Controllers/ViewController.swift
Comment (imaginaris) [reply Example/Controllers/ViewController.swift:142]: You are right. The URL constructor accepts almost everything as string

File: MiniApp/Classes/MiniApp.swift
Comment (imaginaris) [reply MiniApp/Classes/MiniApp.swift:109]: I thought about that. I'd have to pass the handler all the way to `MiniAppWebView`. That'll complicate the code a little bit but I can do that

Comment (ts-leojoseph-r) [reply MiniApp/Classes/MiniApp.swift:109]: Yeah, I know. This looks cleaner, but we have to let the host app know if something goes wrong

File: Example/Controllers/ViewController.swift
Comment (imaginaris) [reply Example/Controllers/ViewController.swift:57]: The MiniApp View can work with miniAppInfo as nil and that is the case during url loading. Adding some special flag to enable miniAppInfo doesn't make sense for me. Plus there is no error handling implemented (it just doesn't display the view)

File: Example/Info.plist
Comment (imaginaris) [reply Example/Info.plist:64]: but it's just for localhost and ip based urls :)

File: Example/Controllers/ViewController.swift
Comment (imaginaris) [reply Example/Controllers/ViewController.swift:142]: error handling added

File: Example/Info.plist
Comment (Climbatize) [reply Example/Info.plist:64]: don't worry there's no way to do differently

File: MiniApp/Classes/Display/MiniAppExternalWebViewController.swift
Comment (Climbatize) [reply MiniApp/Classes/Display/MiniAppExternalWebViewController.swift:12]: Oh I see. But you could add an optional parameter with default value to nil, this way the method remains backward compatible. I really want to get rid of this static member. What do you think @leorjoseph and @corycaywood ?

Comment (imaginaris) [reply MiniApp/Classes/Display/MiniAppExternalWebViewController.swift:12]: I agree. I don't like this static var either. If I get a green light to add optional parameter to the public api method then I'll fix this

File: MiniApp/Classes/MiniApp.swift
Comment (ts-leojoseph-r) [inline MiniApp/Classes/MiniApp.swift:98]: mini URL ? Or self-hosted url ?

**Review by ts-leojoseph-r — COMMENTED** [review]
Thanks for the changes. Overall looks ok to me. Looks like Codecov is failing, please update the test cases for the same.

File: MiniApp/Classes/Display/MiniAppExternalWebViewController.swift
Comment (imaginaris) [reply MiniApp/Classes/Display/MiniAppExternalWebViewController.swift]: Ok, I replaced that static var with an additional parameter in the api without breaking it. Let me know if you like it :)

**Comment by imaginaris** [comment]
My tests are failing on Travis for some reason. They are working fine locally on Xcode 12.2 and 11.7 (tried to change with xcode 11.7 on Travis but that didn't help). Do you know what might be the issue?

**Review by corycaywood — APPROVED** [review]
LGTM 👍  But let's wait for @leorjoseph to approve.

File: fastlane/Fastfile
Comment (ts-leojoseph-r) [inline fastlane/Fastfile:28]: Any reason for this change?

**Review by ts-leojoseph-r — APPROVED** [review]
Sorry, I was off yesterday. Strange to see the test cases failing however the updated test cases will do for now.

Looks good to me :+1:

File: fastlane/Fastfile
Comment (imaginaris) [reply fastlane/Fastfile:28]: iPhone X is not available since Xcode 11, iPhone 8 is more universal (and the ci was picking it as it couldn't find iPhoneX simulator)

Comment (ts-leojoseph-r) [reply fastlane/Fastfile:28]: Ok. Makes sense, just wanna know.

Comment (imaginaris) [reply fastlane/Fastfile]: Ok interesting. I changed this setting back to iPhoneX and the tests started to work. It looks like Travis is creating iPhoneX simulators up to iOS 12.4 so it was always picking this one before. With the iPhone 8, Travis used iOS 13.2 simulator and that seems to cause the issue (upgrading to Xcode 11.7 doesn't help). Anyway I'm leaving the old setting and letting you know, maybe it's some iOS 13.2 specific bug...

**Pull request merged**
